python/cuxfilter/layouts/layouts.py

```diff
@@ -25,234 +13,185 @@ def is_widget(obj):
+    def generate_dashboard(self, title, charts, theme, layout_array=None):
+        pn.config.sizing_mode = "stretch_both"
+        self._layout_array = layout_array
+        if theme == cuxfilter.themes.dark or theme == cuxfilter.themes.rapids:
+            tmpl = ReactTemplate(title=title, theme=DarkTheme, compact="both")
+        else:
+            tmpl = ReactTemplate(title=title, compact="both")
 
         widgets = [x for x in charts.values() if is_widget(x)]
-        plots = [x for x in charts.values() if not is_widget(x)]
+        tmpl = self._process_widgets(widgets, tmpl)
         self._apply_themes(charts, theme)
-        widgetbox = self._process_widgets(widgets)
+        plots = [x for x in charts.values() if not is_widget(x)]
         self._process_plots(plots, tmpl)
-        self._pad_missing_plots(len(plots), tmpl)
-        tmpl.add_panel("widgets", widgetbox)
         return tmpl
 
-    @property
-    def total_charts(self):
-        # Would be nice if the templates and this value could be derived from
-        # some common declarative source description
-        return len(self._num_charts_pat.findall(self._layout))
-
     def _apply_themes(self, charts, theme):
         for chart in charts.values():
             if hasattr(chart, "apply_theme"):
                 chart.apply_theme(theme.chart_properties)
 
-    def _process_widgets(self, widgets_list):
-        widgets_ = pn.Column()
+    def _process_widgets(self, widgets_list, tmpl):
         for obj in widgets_list:
             obj.chart.width = 280
             obj.chart.sizing_mode = "scale_both"
-            widgets_.append(obj.view())
-        return widgets_
-
-    def _pad_missing_plots(self, num_charts_added, tmpl):
-        N = self.total_charts - num_charts_added
+            tmpl.sidebar.append(obj.view())
+        return tmpl
 
-        for i in range(N):
-            tmpl.add_panel(f"chart{self.total_charts - i}", "")
+    def _process_grid_matrix(self, plots, tmpl):
+        arr = np.array(self._layout_array)
+        if len(arr.shape) == 1:
+            arr = np.array([arr])
+        for i in range(arr.max()):
+            if i < len(plots):
+                x0, y0 = (
+                    np.array(
+                        [
+                            np.where(arr == i + 1)[0][0],
+                            np.where(arr == i + 1)[1][0],
+                        ]
+                    )
+                    / np.array(arr.shape)
+                    * (6, 12)
+                )
+                x1, y1 = (
+                    np.array(
+                        [
+                            np.where(arr == i + 1)[0][-1] + 1,
+                            np.where(arr == i + 1)[1][-1] + 1,
+                        ]
+                    )
+                    / np.array(arr.shape)
+                    * (6, 12)
+                )
+                tmpl.main[int(x0) : int(x1), int(y0) : int(y1)] = plots[
+                    i
+                ].view()
```

Comment:
suggest defining a helper to condense these codepaths, e.g similar to
```
def _helper(array, i, pos):
    x, y = np.array([
        np.where(arr == i + 1)[0][pos],
        np.where(arr == i + 1)[1][pos],
    ]) / np.array(arr.shape) * (6, 12)
    return int(x), int(y)
```

Then the above can be streamlined
```
    if i < len(plots):
        x0, y0 = _helper(arr, i, 0)
        x1, y1 = _helper(arr, i, -1) 
        tmpl.main[x0:x1, y0:y1] = plots[i].view()
```

Reply:
thanks, implemented this